data_model/src/lib.rs

```diff
@@ -864,6 +864,83 @@ mod model {
         /// in the next request to continue fetching results of the original query
         pub cursor: crate::query::cursor::ForwardCursor,
     }
+
+    /// String containing serialized valid JSON.
+    ///
+    /// This string is guaranteed to be parsed as JSON.
+    #[derive(Display, Default, Debug, Clone, Eq, Encode, Decode, Ord, PartialOrd, IntoSchema)]
+    #[ffi_type(unsafe {robust})]
+    #[repr(transparent)]
+    #[display(fmt = "{}", "0")]
+    pub struct JsonString(pub(super) String);
+}
+
+impl JsonString {
+    /// Deserialize JSON into something
+    /// # Errors
+    /// See [`serde_json::from_str`].
+    pub fn deserialize<'a, T>(&'a self) -> serde_json::Result<T>
+    where
+        T: Deserialize<'a>,
+    {
+        serde_json::from_str(&self.0)
+    }
+
+    /// Serializes a value into [`JsonString`]
+    /// # Errors
+    /// See [`serde_json::to_string`].
+    pub fn serialize<T: serde::Serialize>(value: &T) -> serde_json::Result<Self> {
+        let serialized = serde_json::to_string(value)?;
+        // the string was obtained from serde_json serialization,
+        // so it should be a valid JSON string
+        Ok(Self(serialized))
+    }
+
+    /// Create without checking whether the input is a valid JSON string.
+    ///
+    /// The caller must guarantee that the value is valid.
+    pub fn from_json_string_unchecked(value: String) -> Self {
+        Self(value)
+    }
+}
+
+impl From<&serde_json::Value> for JsonString {
+    fn from(value: &serde_json::Value) -> Self {
+        Self(value.to_string())
+    }
+}
+
+impl From<serde_json::Value> for JsonString {
+    fn from(value: serde_json::Value) -> Self {
+        Self::from(&value)
+    }
+}
+
+impl PartialEq for JsonString {
```

Comment:
it's ok to derive this

Reply:
look at docs for `Permission` about being compared by payload and copy that here on `JsonString`

Reply:
as you wish actually, but then you have to make `PartialOrd` consistent with `PartialEq`. Just make sure to transfer the docs here

Reply:
the thing is that `asset##account@domain` will compare as not equal to `asset#domain#account@domain` so I thought we may as well not care about whitespaces